File: farm_management_system/savanna_farm_suite/report/cow_production_report/cow_production_report.py
# Copyright (c) 2025, Techsavanna Technology and contributors
# For license information, please see license.txt

# import frappe
from __future__ import annotations
import frappe
from frappe.utils import getdate, today
from datetime import date, timedelta

def get_date_range_for_timeline(timeline: str):
    td_today = getdate(today())
    if timeline == "This Week":
        start = td_today - timedelta(days=td_today.weekday())
        end = td_today
    elif timeline == "Last Fortnight":
        start = td_today - timedelta(days=13)
        end = td_today
    elif timeline == "This Month":
        start = date(td_today.year, td_today.month, 1)
        end = td_today
    elif timeline == "This Quarter":
        quarter_index = (td_today.month - 1) // 3
        start_month = quarter_index * 3 + 1
        start = date(td_today.year, start_month, 1)
        end = td_today
    elif timeline == "This Year":
        start = date(td_today.year, 1, 1)
        end = td_today
    else:
        start = date(td_today.year, td_today.month, 1)
        end = td_today
    return start, end

def safe_get_current_stock(product_name):
    """
    Best-effort wrapper around your app's get_current_stock.
    Returns dict with keys: qty_after_transaction, stock_value, valuation_rate (or None on failure).
    """
    try:
        fn = frappe.get_attr("farm_management_system.savanna_farm_suite.doctype.crop_seedlings.crop_seedlings.get_current_stock")
        res = fn(product_name)
        if isinstance(res, dict):
            return {
                "qty_after_transaction": res.get("qty_after_transaction"),
                "stock_value": res.get("stock_value"),
                "valuation_rate": res.get("valuation_rate")
            }
    except Exception:
        pass
    return {"qty_after_transaction": None, "stock_value": None, "valuation_rate": None}

def execute(filters=None):
    if filters is None:
        filters = {}

    columns = [
        "Cow:Link/Cattle:180",
        "Cow Nickname:Data:140",
        "Date of Collection:Date:120",
        "Product Collected:Link/Animal Products:220",
        "Quantity Collected:Float:120",
        "UOM:Data:80",
        "Stock Value:Currency:120",
        "Selling Rate:Currency:120"
    ]

    timeline = (filters.get("timeline") or "").strip() or "This Month"
    start_date, end_date = get_date_range_for_timeline(timeline)

    cow_filter = filters.get("cow")
    product_filter = filters.get("product")

    # aggregation: key = (cow_name, iso_date, product_collected, uom) -> total_qty
    agg = {}

    # Preload list of cattle (narrow by cow_filter if provided)
    if cow_filter:
        cattle_list = frappe.get_all("Cattle", filters={"name": cow_filter}, fields=["name"])
    else:
        cattle_list = frappe.get_all("Cattle", fields=["name"])

    # We'll cache nicknames per cattle to avoid repeated fetches later
    nickname_map = {}

    for c in cattle_list:
        try:
            cattle_doc = frappe.get_doc("Cattle", c.name)
        except Exception:
            continue

        # cache nickname
        nickname_map[c.name] = (cattle_doc.get("add_nickname_optional") or "") if hasattr(cattle_doc, "get") else (getattr(cattle_doc, "add_nickname_optional", "") or "")

        rows = getattr(cattle_doc, "production_log", []) or []
        for row in rows:
            # support both dict-like and object child rows
            if isinstance(row, dict):
                row_date = row.get("date_of_collection")
                prod = row.get("product_collected")
                qty = row.get("quantity_collected", 0)
                uom = (row.get("products_default_uom") or "")
            else:
                row_date = getattr(row, "date_of_collection", None)
                prod = getattr(row, "product_collected", None)
                qty = getattr(row, "quantity_collected", 0)
                uom = (getattr(row, "products_default_uom", "") or "")

            if not row_date:
                continue

            try:
                d = getdate(row_date)
            except Exception:
                continue

            # date range filter (inclusive)
            if d < start_date or d > end_date:
                continue

            # product filter
            if product_filter and prod != product_filter:
                continue

            try:
                qty_f = float(qty or 0)
            except Exception:
                # skip rows with non-numeric qty
                continue

            key = (c.name, d.isoformat(), prod or "", uom or "")
            agg[key] = agg.get(key, 0.0) + qty_f

    # Build final result rows
    data = []
    sorted_keys = sorted(agg.keys(), key=lambda k: (k[0] or "", k[1] or ""))

    for key in sorted_keys:
        cow_name, iso_date, product_collected, uom = key
        total_qty = agg[key]

        # fetch stock value & valuation_rate (best effort)
        stock_info = safe_get_current_stock(product_collected)
        stock_value = stock_info.get("stock_value")
        valuation_rate = stock_info.get("valuation_rate")

        # Prepare numeric values (float) for report columns
        try:
            qty_value = float(total_qty)
        except Exception:
            qty_value = 0.0

        # Stock and valuation: ensure floats or None (Frappe will render blank for None)
        try:
            sv_value = float(stock_value) if stock_value is not None else None
        except Exception:
            sv_value = None

        try:
            vr_value = float(valuation_rate) if valuation_rate is not None else None
        except Exception:
            vr_value = None

        # Cow nickname from cache (may be empty string)
        nickname = nickname_map.get(cow_name, "")

        # Append row in the same order as columns
        data.append([
            cow_name,          # Cow (Link)
            nickname,          # Cow Nickname (Data)
            iso_date,          # Date of Collection (Date string iso)
            product_collected, # Product Collected (Link)
            qty_value,         # Quantity Collected (Float numeric)
            uom or "",         # UOM (Data)
            sv_value,          # Stock Value (Currency numeric or None)
            vr_value           # Valuation Rate (Currency numeric or None)
        ])

    return columns, data

File: farm_management_system/savanna_farm_suite/report/cow_production_report/cow_production_report.js
// Copyright (c) 2025, Techsavanna Technology and contributors
// For license information, please see license.txt

frappe.query_reports["Cow Production Report"] = {
	"filters": [
		{
			"fieldname": "cow",
			"label": __("Narrow Productivity to a cow"),
			"fieldtype": "Link",
			"options": "Cattle",
			"reqd": 0
		},
		{
			"fieldname": "product",
			"label": __("Specify Animal Product"),
			"fieldtype": "Link",
			"options": "Animal Products",
			"reqd": 0,
			"get_query": function() {
				return {
					filters: {
						"product_tied_to_which_animal": "Cow"
					}
				};
			}
		},
		{
			"fieldname": "timeline",
			"label": __("Specify Timeline"),
			"fieldtype": "Select",
			"options": "\nThis Week\nLast Fortnight\nThis Month\nThis Quarter\nThis Year",
			"default": "This Month",
			"reqd": 1
		}
	],

	// create DOM holder for chart on first load
	"onload": function(report) {
		// add container above the report results
		if (!report.page.page_chart_area) {
			const $chart_wrapper = $('<div class="cow-production-chart-wrapper" style="margin-bottom: 20px;"></div>');
			report.page.page_chart_area = $chart_wrapper;
			report.page.main.prepend($chart_wrapper);
		}
	},

	// // on refresh, re-run the report quietly to get data and draw chart
	// "on_refresh": function(report, dt, filters) {
	// 	// build filters object
	// 	const f = report.get_values ? report.get_values() : (report.page_fields || {}).get_values ? report.page_fields.get_values() : filters;

	// 	// call the query backend to get the same data as the table
	// 	frappe.call({
	// 		method: 'frappe.desk.query_report.run',
	// 		args: {
	// 			report_name: 'Cow Production Report',
	// 			filters: f || {}
	// 		},
	// 		callback: function(res) {
	// 			if (!res || !res.message) return;
	// 			const msg = res.message;
	// 			const rows = msg.result || [];
	// 			const $chart_wrapper = report.page.page_chart_area;
	// 			$chart_wrapper.empty();

	// 			if (!rows.length) {
	// 				$chart_wrapper.append($('<div class="text-muted" style="padding: 15px;">No data for chart.</div>'));
	// 				return;
	// 			}

	// 			// --- Chart Data Processing ---

	// 			// 1. Find column indexes dynamically
	// 			const colNames = (msg.columns || []).map(c => (typeof c === 'string' ? c.split(':')[0].trim() : (c.label || '')));
	// 			const idxCow = colNames.indexOf('Cow');
	// 			const idxProduct = colNames.indexOf('Product Collected');
	// 			const idxQty = colNames.indexOf('Quantity Collected');

	// 			// 2. Aggregate data for a stacked bar chart
	// 			// We need a structure like: { cowName: { productName: totalQty, ... }, ... }
	// 			const dataMap = {};
	// 			const allProducts = new Set();
	// 			const allCows = new Set();

	// 			rows.forEach(row => {
	// 				const cow = row[idxCow] || 'Unknown';
	// 				const product = row[idxProduct] || 'Unnamed Product';
	// 				const qtyCell = (row[idxQty] || '0').toString();
	// 				const match = qtyCell.match(/-?\d+(\.\d+)?/);
	// 				const qty = match ? parseFloat(match[0]) : 0;

	// 				if (!dataMap[cow]) {
	// 					dataMap[cow] = {};
	// 				}
	// 				dataMap[cow][product] = (dataMap[cow][product] || 0) + qty;
					
	// 				allCows.add(cow);
	// 				allProducts.add(product);
	// 			});

	// 			// 3. Prepare data in Frappe Charts format
	// 			const labels = Array.from(allCows).sort(); // X-axis labels (Cow names)
	// 			const uniqueProducts = Array.from(allProducts).sort();

	// 			const datasets = uniqueProducts.map(product => {
	// 				return {
	// 					name: product,
	// 					values: labels.map(cow => dataMap[cow][product] || 0) // Get qty for this product for each cow
	// 				};
	// 			});

	// 			// 4. Generate a consistent random color for each product
	// 			const productColors = {};
	// 			const randomColor = () => '#' + Math.floor(Math.random()*16777215).toString(16).padStart(6, '0');
	// 			uniqueProducts.forEach(p => {
	// 				productColors[p] = randomColor();
	// 			});
	// 			const colors = uniqueProducts.map(p => productColors[p]);

	// 			// --- Render Chart ---
	// 			const chart_id = 'cow-production-chart';
	// 			$chart_wrapper.append(`<div id="${chart_id}"></div>`);

	// 			new frappe.Chart(`#${chart_id}`, {
	// 				title: __('Cow Production by Product'),
	// 				data: {
	// 					labels: labels,
	// 					datasets: datasets
	// 				},
	// 				type: 'bar', // bar chart
	// 				height: 350,
	// 				colors: colors,
	// 				stacked: true, // This creates the stacked bar effect!
	// 				tooltipOptions: {
	// 					formatTooltipY: d => d + ' units'
	// 				}
	// 			});
	// 		}
	// 	});
	// }
};